neqo-transport/src/connection.rs

```diff
@@ -1770,6 +1804,49 @@ mod tests {
         assert_eq!(*server.state(), State::Connected);
     }
 
+    #[test]
+    fn test_conn_handshake_failed_authentication() {
+        qdebug!("---- client: generate CH");
+        let mut client = default_client();
+        let out = client.process(None, now());
+        assert!(out.as_dgram_ref().is_some());
+        qdebug!("Output={:0x?}", out.as_dgram_ref());
+
+        qdebug!("---- server: CH -> SH, EE, CERT, CV, FIN");
+        let mut server = default_server();
+        let out = server.process(out.dgram(), now());
+        assert!(out.as_dgram_ref().is_some());
+        qdebug!("Output={:0x?}", out.as_dgram_ref());
+
+        qdebug!("---- client: cert verification");
+        let out = client.process(out.dgram(), now());
+        assert!(out.as_dgram_ref().is_some());
+        qdebug!("Output={:0x?}", out.as_dgram_ref());
+
+        let out = server.process(out.dgram(), now());
+        assert!(out.as_dgram_ref().is_none());
+        qdebug!("Output={:0x?}", out.as_dgram_ref());
+
+        let authentication_needed = |e| matches!(e, ConnectionEvent::AuthenticationNeeded);
+        assert!(client.events().any(authentication_needed));
+        qdebug!("---- client: Alert(certificate_revoked)");
+        client.authenticated(-(0x2000) + 12, now());
```

Comment:
This is where I think that this API is a little troublesome.  Getting the right `PRErrorCode` value seems to break encapsulation badly.  Maybe don't fix that now, but open an issue to track passing an enum in here instead.

Reply:
I will open another issue.

Reply:
https://github.com/mozilla/neqo/issues/142